src/components/common/BaseDatasourceSettings.tsx

```diff
@@ -0,0 +1,109 @@
+import * as React from 'react';
+import * as _ from 'lodash';
+
+import Card from 'react-md/lib/Cards/Card';
+import TextField from 'react-md/lib/TextFields';
+import FontIcon from 'react-md/lib/FontIcons';
+import CardTitle from 'react-md/lib/Cards/CardTitle';
+import Switch from 'react-md/lib/SelectionControls/Switch';
+import Avatar from 'react-md/lib/Avatars';
+import SelectField from 'react-md/lib/SelectFields';
+
+import SettingsActions from '../../actions/SettingsActions';
+
+export interface IBaseSettingsProps {
+  settings: DataSource;
+}
+
+export interface IBaseSettingsState { 
+}
+
+export abstract class BaseDatasourceSettings<T extends IBaseSettingsState> 
+                      extends React.Component<IBaseSettingsProps, T> {
+
+  // require derived classes to implement
+  abstract icon: string;
+  abstract renderChildren();
+  
+  constructor(props: IBaseSettingsProps) {
+    super(props);
+
+    this.onParamChange = this.onParamChange.bind(this);
+    this.onParamSelectChange = this.onParamSelectChange.bind(this);
+    this.getProperty = this.getProperty.bind(this);
+    this.updateProperty = this.updateProperty.bind(this);
+    this.renderChildren = this.renderChildren.bind(this);
+  }
+
+  protected getProperty(property: string, defaultValue: any = null): any {
```

Comment:
Add some comments on the methods in this class (and the class itself) to explain their usage